Detectcone cleanup

arbulgazar opened the pull request:
* Removes unnacisaritly files
* Improves structure
* Adds automatic downloads of cluon and eigen lib
* Updates for the use of cluon-v114

arbulgazar committed:
Repository refactoring and cleanup

* Removes unused cluons
* Removes static dependancy on Eigen, download fresh from git
* Moves header files to include dir

arbulgazar committed:
Adds automatic cluon get from server

Updates to latest cluon-v114

arbulgazar committed:
Removes static cluon header

chrberger commented on CMakeLists.txt:
```diff
@@ -55,19 +55,21 @@ set(THREADS_PREFER_PTHREAD_FLAG ON)
 ################################################################################
 # Extract cluon-msc from cluon-complete.hpp.
+
+execute_process(COMMAND wget -P ${CLUON_PATH} https://raw.githubusercontent.com/chrberger/libcluon/gh-pages/headeronly/${CLUON_COMPLETE})
```
@arbulgazar Is this line executed once or on any build?

arbulgazar replied:
@chrberger yes, executed on every build, but if the file is already there it is not downloaded again. Should I make this check explicitly?

chrberger replied:
The issue might be that if you want to build when you only have the repository at hand and no Internet connection, this step might fail?

arbulgazar replied:
Regardless - will add an explicit check, will be cleaner.

bjornborg replied:
If you run this outside of the docker, e.g. natively, it will clutter the repository.

chrberger commented on include/detectcone.hpp:
```diff
@@ -17,127 +17,125 @@
-  
-
-
+  const double PI = 3.14159265;
```
@arbulgazar We should use the constant provided in #include <cmath> (M_PI)

arbulgazar replied:
What about `DEG2RAD` and `RAD2DEG`?

chrberger replied:
Should be adjusted in the same way (including `constexpr` to let the compile provide these constants even during compile time for further optimizations)

chrberger reviewed (changes requested):
The file https://github.com/chalmersfsd/cfsd18-perception-detectcone/blob/detectcone_cleanup/docker-compose.yml contains:
* a specific user local path
* the use of libcluon's POSIX shared memory implementation (which is faster than libcluon's SystemV IPC's shared memory implementation); the consequence, though, is that all microservices that belong to a shared memory processing chain (i.e., producers, consumers) *must* be using the *same* Docker base image (i.e., all *must* use Alpine Linux or all *must* use Ubuntu); mixing different Docker base images result in a not working processing chain
* the file `model` is empty
* the README.md file must be updated to contain information how the service is supposed to be used
* does Travis still complete successfully?
* there is no valid test case: https://github.com/chalmersfsd/cfsd18-perception-detectcone/blob/detectcone_cleanup/test/tests-cfsd18-perception-detectcone.cpp

bjornborg commented on CMakeLists.txt:
```diff
@@ -55,39 +55,41 @@ set(THREADS_PREFER_PTHREAD_FLAG ON)
 
 # Add current build directory as include directory as it contains generated files.
 include_directories(SYSTEM ${CMAKE_BINARY_DIR})
 include_directories(SYSTEM ${CMAKE_CURRENT_SOURCE_DIR}/thirdparty)
```
This will give you a warning/error when compiling since the directory no longer exists.
`[ 45%] Building CXX object CMakeFiles/cfsd18-perception-detectcone-core.dir/src/collector.cpp.o
cc1plus: error: /opt/sources/thirdparty: No such file or directory [-Werror=missing-include-dirs]
cc1plus: error: /opt/sources/thirdparty: No such file or directory [-Werror=missing-include-dirs]
cc1plus: error: /opt/sources/thirdparty: No such file or directory [-Werror=missing-include-dirs]
cc1plus: all warnings being treated as errors
make[2]: *** [CMakeFiles/cfsd18-perception-detectcone-core.dir/build.make:89: CMakeFiles/cfsd18-perception-detectcone-core.dir/src/cone.cpp.o] Error 1
make[2]: *** Waiting for unfinished jobs....
In file included from /usr/local/include/tiny_dnn/xtensor/xcontainer.hpp:17,
                 from /usr/local/include/tiny_dnn/xtensor/xarray.hpp:17,
                 from /usr/local/include/tiny_dnn/util/util.h:24,
                 from /usr/local/include/tiny_dnn/lossfunctions/loss_function.h:12,
                 from /usr/local/include/tiny_dnn/network.h:26,
                 from /usr/local/include/tiny_dnn/tiny_dnn.h:11,
                 from /opt/sources/include/detectcone.hpp:46,
                 from /opt/sources/src/detectcone.cpp:19:
/usr/local/include/tiny_dnn/xtensor/xiterable.hpp: In member function 'xt::xiterable<D>::layout_iterator<L> xt::xiterable<D>::storage_begin()':
/usr/local/include/tiny_dnn/xtensor/xiterable.hpp:740:19: error: expected primary-expression before ')' token
   return begin<L>();
                   ^
/usr/local/include/tiny_dnn/xtensor/xiterable.hpp: In member function 'xt::xiterable<D>::layout_iterator<L> xt::xiterable<D>::storage_end()':
/usr/local/include/tiny_dnn/xtensor/xiterable.hpp:746:17: error: expected primary-expression before ')' token
   return end<L>();
                 ^
/usr/local/include/tiny_dnn/xtensor/xiterable.hpp: In member function 'xt::xiterable<D>::reverse_layout_iterator<L> xt::xiterable<D>::storage_rbegin()':
/usr/local/include/tiny_dnn/xtensor/xiterable.hpp:753:20: error: expected primary-expression before ')' token
   return rbegin<L>();
                    ^
/usr/local/include/tiny_dnn/xtensor/xiterable.hpp: In member function 'xt::xiterable<D>::reverse_layout_iterator<L> xt::xiterable<D>::storage_rend()':
/usr/local/include/tiny_dnn/xtensor/xiterable.hpp:760:18: error: expected primary-expression before ')' token
   return rend<L>();
                  ^
In file included from /usr/local/include/tiny_dnn/xtensor/xcontainer.hpp:17,
                 from /usr/local/include/tiny_dnn/xtensor/xarray.hpp:17,
                 from /usr/local/include/tiny_dnn/util/util.h:24,
                 from /usr/local/include/tiny_dnn/lossfunctions/loss_function.h:12,
                 from /usr/local/include/tiny_dnn/network.h:26,
                 from /usr/local/include/tiny_dnn/tiny_dnn.h:11,
                 from /opt/sources/include/detectcone.hpp:46,
                 from /opt/sources/include/collector.hpp:29,
                 from /opt/sources/src/collector.cpp:20:
/usr/local/include/tiny_dnn/xtensor/xiterable.hpp: In member function 'xt::xiterable<D>::layout_iterator<L> xt::xiterable<D>::storage_begin()':
/usr/local/include/tiny_dnn/xtensor/xiterable.hpp:740:19: error: expected primary-expression before ')' token
   return begin<L>();
                   ^
/usr/local/include/tiny_dnn/xtensor/xiterable.hpp: In member function 'xt::xiterable<D>::layout_iterator<L> xt::xiterable<D>::storage_end()':
/usr/local/include/tiny_dnn/xtensor/xiterable.hpp:746:17: error: expected primary-expression before ')' token
   return end<L>();
                 ^
/usr/local/include/tiny_dnn/xtensor/xiterable.hpp: In member function 'xt::xiterable<D>::reverse_layout_iterator<L> xt::xiterable<D>::storage_rbegin()':
/usr/local/include/tiny_dnn/xtensor/xiterable.hpp:753:20: error: expected primary-expression before ')' token
   return rbegin<L>();
                    ^
/usr/local/include/tiny_dnn/xtensor/xiterable.hpp: In member function 'xt::xiterable<D>::reverse_layout_iterator<L> xt::xiterable<D>::storage_rend()':
/usr/local/include/tiny_dnn/xtensor/xiterable.hpp:760:18: error: expected primary-expression before ')' token
   return rend<L>();
                  ^
/opt/sources/src/detectcone.cpp: In member function 'void DetectCone::forwardDetectionORB(cv::Mat)':
/opt/sources/src/detectcone.cpp:580:38: error: array must be initialized with a brace-enclosed initializer
   cv::Mat probMap[4] = cv::Mat::zeros(m_height, m_width, CV_64F);
                        ~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~
cc1plus: all warnings being treated as errors
make[2]: *** [CMakeFiles/cfsd18-perception-detectcone-core.dir/build.make:76: CMakeFiles/cfsd18-perception-detectcone-core.dir/src/collector.cpp.o] Error 1
cc1plus: all warnings being treated as errors
make[2]: *** [CMakeFiles/cfsd18-perception-detectcone-core.dir/build.make:63: CMakeFiles/cfsd18-perception-detectcone-core.dir/src/detectcone.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:105: CMakeFiles/cfsd18-perception-detectcone-core.dir/all] Error 2
make: *** [Makefile:141: all] Error 2
The command '/bin/sh -c mkdir -p build &&     cd build &&     cmake -D CMAKE_BUILD_TYPE=Release -D CMAKE_INSTALL_PREFIX=/tmp/cfsd18-perception-detectcone-dest .. &&     make -j3 && make test && make install && upx -9 /tmp/cfsd18-perception-detectcone-dest/bin/cfsd18-perception-detectcone' returned a non-zero code: 2
`

bjornborg commented on include/detectcone.hpp:
```diff
@@ -17,127 +17,125 @@
  * USA.
  */
 
-#ifndef OPENDLV_LOGIC_CFSD18_SENSATION_ATTENTION_HPP
```
I am also getting warnings/errors from cv::Mat at detectcone.cpp:580
`/opt/sources/src/detectcone.cpp: In member function 'void DetectCone::forwardDetectionORB(cv::Mat)':
/opt/sources/src/detectcone.cpp:580:38: error: array must be initialized with a brace-enclosed initializer
   cv::Mat probMap[4] = cv::Mat::zeros(m_height, m_width, CV_64F);
                        ~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~`

bjornborg commented on CMakeLists.txt:
```diff
@@ -20,8 +20,8 @@ project(cfsd18-perception-detectcone)
 ################################################################################
```
You might want to change the header author to Cfsd instead of Christian :D @chrberger .

chrberger replied:
Yes, please :-)

bjornborg commented on CMakeLists.txt:
```diff
@@ -20,8 +20,8 @@ project(cfsd18-perception-detectcone)
 ################################################################################
 # Defining the relevant versions of OpenDLV Standard Message Set and libcluon.
```
I usually add: 

find_program(CCACHE_FOUND` ccache)

if(CCACHE_FOUND)
    set_property(GLOBAL PROPERTY RULE_LAUNCH_COMPILE ccache)
    set_property(GLOBAL PROPERTY RULE_LAUNCH_LINK ccache)
endif(CCACHE_FOUND)

to use ccache on the machines where it is installed. It often speeds up local compilation when you want to test things out.

bjornborg reviewed (changes requested):
I think it would be nice with added
- read me with pictures or gif showing the function
- usecase showcasing the functionality with data
- update the travis script

things to be fixed
- Dockerfile

And before we merge the PR, we should also make a version release to highlight that this is latest from cfsd18 @ZiweiHuang94 .

ruguangyou committed:
Add some comments in main function

ruguangyou committed:
Update README.md

ruguangyou committed:
Merge branch 'detectcone_cleanup' of https://github.com/chalmersfsd/cfsd18-perception-detectcone into detectcone_cleanup

ruguangyou committed:
Update README.md, add od4 message input and output

ruguangyou committed:
Update README.md

ruguangyou committed:
Update README.md (cone color labels)

arbulgazar committed:
Fix missing thirdparty dir

bjornborg committed:
Cleaned up some temporary files

Changed the alpine version to an older version for Gcc6. Gcc8 is not compatible with tinydnn

Restructured the repo

The config file is not really needed, the settings can be parsed through cmake command

bjornborg committed:
Removed the config file

bjornborg committed:
Cleaned up cmake module mess, turns out that we didn't need them
Tweaked the eigen include

bjornborg committed:
Disabled upx compression for binary

ruguangyou committed:
slim down by only copying dependent libraries (image size: 393MB -> 79.1MB)

bjornborg committed:
Merge pull request #4 from chalmersfsd/detectcone_cleanup.patch

Detectcone cleanup.patch

bjornborg commented:
Is it possible to use alpine as deployment base image on arch @chrberger ?

bjornborg commented:
Left to do:

- [ ] Update travis script
- [ ] Update version control
- [ ] Introduce unit tests for the module

chrberger commented:
@bjornborg "Is it possible to use Alpine?"

Yes, when the following constraints are met:
a) the producing microservice that is giving us the image via a shared memory is also Alpine
b) when the producing microservice is not Alpine but the environment variable `CLUON_SHAREDMEMORY_POSIX` is either not set or set to 0 (cf. here: https://github.com/chrberger/libcluon/blob/master/libcluon/src/SharedMemory.cpp#L59-L61).

Reason: When the producer is non-Alpine and the receiver (ie., this microservice) is Alpine, the C-libraries are incompatible and hence, are not allowing using the fast POSIX shared memory implementation but only the (slightly) slower SysV IPC implementation.
